### PR TITLE
Make sync worker only run when needed

### PR DESCRIPTION
**Pull request by JonasMayerDev**
1. Before running sync worker, check if another sync worker is already running.
Before in theory 2 sync jobs could run and potentially mess with each other (maybe leads to https://github.com/nextcloud/android/issues/11974)
2. Check if any sync folder is enabled and early exit if not.
3. Only scan the file system if "only charging" is disabled or is charging

- [X] Tests written, or not not needed

### CODE REVIEW

**Review by ZetaTom — APPROVED** [review]
In my brief testing everything worked as expected. Although I didn't check any edge cases (what happens when a worker is stuck, for example). This might be out of scope for this pull request.

The suggestions are just a minor change and, in reality, just eye candy.

File: app/src/main/java/com/nextcloud/utils/extensions/WorkManagerExtensions.kt [hunk 1]
```diff
@@ -47,3 +47,24 @@ fun WorkManager.isWorkScheduled(tag: String): Boolean {
...
+    } catch (e: InterruptedException) {
+        Log_OC.d("Worker", "InterruptedException in isWorkScheduled: $e")
+    }
+
+    for (workInfo in workInfoList) {
+        val state = workInfo.state
+        running = running || state == WorkInfo.State.RUNNING
+    }
+
+    return running
```
[the remark this answers is not in the document]
Comment (JonasMayerDev) [reply app/src/main/java/com/nextcloud/utils/extensions/WorkManagerExtensions.kt]: Thanks, I have replaced 2 for loops with your solution :)

**Comment by JonasMayerDev** [comment]
Problem: With current implementation when periodic sync worker is started and immediate filesync is already running because of changed files it will skip periodic sync which is not intended... But it is really unlikely since immediate file sync with changed files should only run for under 1 sec.

**Review by alperozturk96 — CHANGES_REQUESTED** [review]
We can get rid of repeated codes in WorkManagerExtensions with this code.

```
private const val TAG = "WorkManager"

fun WorkManager.isWorkRunning(tag: String): Boolean = checkWork(tag, listOf(WorkInfo.State.RUNNING))

fun WorkManager.isWorkScheduled(tag: String): Boolean =
    checkWork(tag, listOf(WorkInfo.State.RUNNING, WorkInfo.State.ENQUEUED))

private fun WorkManager.checkWork(tag: String, stateConditions: List<WorkInfo.State>): Boolean {
    val statuses: ListenableFuture<List<WorkInfo>> = getWorkInfosByTag(tag)
    var workInfoList: List<WorkInfo> = emptyList()

    try {
        workInfoList = statuses.get()
    } catch (e: ExecutionException) {
        Log_OC.d(TAG, "ExecutionException in checkWork: $e")
    } catch (e: InterruptedException) {
        Log_OC.d(TAG, "InterruptedException in checkWork: $e")
    }

    return workInfoList.any { workInfo -> stateConditions.contains(workInfo.state) }
}

```

**Review by alperozturk96 — APPROVED** [review]
Tested didn't came across with any regression
